x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java

```diff
@@ -93,34 +105,126 @@ protected void readInternal(byte[] b, int offset, int length) throws IOException
+                position += read;
+                pos += read;
+                offset += read;
+                length -= read;
+
+                if (streamForSequentialReads.isFullyRead()) {
+                    if (streamForSequentialReadsRef.compareAndSet(streamForSequentialReads, null) != false) {
+                        streamForSequentialReads.close();
+                    } else {
+                        // something happened concurrently, defensively stop optimizing
+                        sequentialReadSize = NO_SEQUENTIAL_READ_OPTIMIZATION;
+                    }
+
+                    if (length == 0) {
+                        // the current stream contained precisely enough data for this read, so we're good.
+                        return;
+                    } else {
+                        // the current stream didn't contain enough data for this read, so we must read more
+                        if (sequentialReadSize != NO_SEQUENTIAL_READ_OPTIMIZATION
+                            && tryReadAndKeepStreamOpen(part, pos, b, offset, length, currentSequentialReadSize)) {
+                            return;
+                        }
+                    }
+                } else {
+                    // the current stream contained enough data for this read and more besides, so we leave it alone.
+                    assert length == 0 : length + " remaining";
+                    return;
+                }
+            } else {
+                // not a sequential read, so stop optimizing for this usage pattern and fall through to the unoptimized behaviour
+                assert streamForSequentialReads.isFullyRead() == false;
+                sequentialReadSize = NO_SEQUENTIAL_READ_OPTIMIZATION;
+                IOUtils.close(streamForSequentialReadsRef.getAndSet(null));
+            }
+        }
+
+        // read part of a blob directly; the code above falls through to this case where there is no optimization possible
         try (InputStream inputStream = blobContainer.readBlob(fileInfo.partName(part), pos, length)) {
-            int read = inputStream.read(b, offset, length);
-            assert read == length;
+            final int read = inputStream.read(b, offset, length);
+            assert read == length : read + " vs " + length;
             position += read;
         }
     }
 
+    /**
+     * If appropriate, open a new stream for sequential reading and satisfy the given read using it. Returns whether this happened or not;
+     * if it did not happen then nothing was read, and the caller should perform the read directly.
+     */
+    private boolean tryReadAndKeepStreamOpen(int part, long pos, byte[] b, int offset, int length, long currentSequentialReadSize)
+        throws IOException {
+
+        assert streamForSequentialReadsRef.get() == null : "should only be called when a new stream is needed";
+        assert currentSequentialReadSize > 0L : "should not be called if ";
```

Comment:
something missing here

Reply:
Fixed in 78087ab4e1c08c13002abfa5a37a1606cd4d7f98.